### PR TITLE
Node admission/authorization 

### PR DESCRIPTION
**Pull request by liggitt**
Enables the node authorizer and NodeRestriction admission plugins

### CODE REVIEW

**Comment by liggitt** [comment]
[test]

**Comment by liggitt** [comment]
Will wait for the 1.7 rebase and drop all the upstream commits

**Comment by liggitt** [comment]
cc @openshift/sig-security

File: pkg/cmd/server/bootstrappolicy/policy.go [hunk 1]
```diff
@@ -966,9 +966,15 @@ func GetOpenshiftBootstrapClusterRoleBindings() []rbac.ClusterRoleBinding {
...
-			BindingOrDie(),
+		// Preserve the empty node binding with no subjects for tightening reconciliation
+		rbac.ClusterRoleBinding{
+			ObjectMeta: metav1.ObjectMeta{Name: NodeRoleBindingName},
+			RoleRef: rbac.RoleRef{
+				APIGroup: rbac.GroupName,
+				Kind:     "ClusterRole",
+				Name:     NodeRoleName,
+			},
+		},
```
Comment (simo5) [inline pkg/cmd/server/bootstrappolicy/policy.go]: please use newOriginClusterBinding() or even better rbac.newClusterBinding() if you do not have to maintain a name difference between the role and the binding.

File: pkg/cmd/server/bootstrappolicy/policy_test.go [hunk 2]
```diff
@@ -176,4 +180,19 @@ func TestCovers(t *testing.T) {
...
+			t.Logf("\t%s", r.CompactString())
+		}
+	}
+	// Make sure our node role doesn't have any extra permissions
+	if covers, miss := rulevalidation.Covers(kbootstrappolicy.NodeRules(), nodeRole.Rules); !covers {
+		t.Errorf("openshift node role has extra permissions:")
+		for _, r := range miss {
+			t.Logf("\t%s", r.CompactString())
+		}
+	}
```
Comment (simo5) [inline pkg/cmd/server/bootstrappolicy/policy_test.go]: Why is this needed ?

File: pkg/cmd/server/bootstrappolicy/policy.go (hunk 1 above)
Comment (liggitt) [reply pkg/cmd/server/bootstrappolicy/policy.go]: that helper does not allow empty subjects, which we want here.

Comment (simo5) [reply pkg/cmd/server/bootstrappolicy/policy.go]: very annoying, what is the purpose of a binding with no subjects ?

Comment (liggitt) [reply pkg/cmd/server/bootstrappolicy/policy.go]: so that tightening reconciliation removes the grant to the `system:nodes` group

File: pkg/cmd/server/origin/master_config.go [hunk 3]
```diff
@@ -779,14 +777,25 @@ func buildKubeAuth(r rbacinformers.Interface) (kauthorizer.Authorizer, rbacregis
 	return kubeAuthorizer, ruleResolver, kubeSubjectLocator
 }
 
-func newAuthorizer(kubeAuthorizer kauthorizer.Authorizer, kubeSubjectLocator rbacauthorizer.SubjectLocator, clusterRoleGetter rbaclisters.ClusterRoleLister, projectRequestDenyMessage string) (kauthorizer.Authorizer, authorizer.SubjectLocator) {
+func newAuthorizer(kubeAuthorizer kauthorizer.Authorizer, kubeSubjectLocator rbacauthorizer.SubjectLocator, informers InformerAccess, projectRequestDenyMessage string) (kauthorizer.Authorizer, authorizer.SubjectLocator) {
```
Comment (deads2k) [inline pkg/cmd/server/origin/master_config.go]: We're trying to kill the big one.  Plumb through the ones you want or make this a method on the config type or make a `Complete` method and we can try to start converging.

Comment (liggitt) [reply pkg/cmd/server/origin/master_config.go]: so plumbed

**Comment by liggitt** [comment]
Router flake
/retest

**Comment by liggitt** [comment]
any other comments?

File: pkg/cmd/server/bootstrappolicy/policy.go [hunk 4]
```diff
@@ -966,9 +966,15 @@ func GetOpenshiftBootstrapClusterRoleBindings() []rbac.ClusterRoleBinding {
 		newOriginClusterBinding(StatusCheckerRoleBindingName, StatusCheckerRoleName).
 			Groups(AuthenticatedGroup, UnauthenticatedGroup).
 			BindingOrDie(),
-		newOriginClusterBinding(NodeRoleBindingName, NodeRoleName).
-			Groups(NodesGroup).
-			BindingOrDie(),
+		// Preserve the empty node binding with no subjects for tightening reconciliation
```
Comment (enj) [inline pkg/cmd/server/bootstrappolicy/policy.go]: Seems like we should have a `GetDeadClusterRoleBindings`.

Comment (liggitt) [reply pkg/cmd/server/bootstrappolicy/policy.go]: I wouldn't want to chance someone calling this and missing calling that

Comment (enj) [reply pkg/cmd/server/bootstrappolicy/policy.go]: We already trust callers to use `GetBootstrapClusterRoles` which gives them `GetDeadClusterRoles`.  I would expect `GetBootstrapClusterRoleBindings` to have the same logic.

Comment (liggitt) [reply pkg/cmd/server/bootstrappolicy/policy.go]: ah, ok

Comment (liggitt) [reply pkg/cmd/server/bootstrappolicy/policy.go]: created GetDeadClusterRoleBindings

**Comment by liggitt** [comment]
updated, and added an integration test that replicates the checks in https://github.com/kubernetes/kubernetes/blob/release-1.7/test/integration/auth/node_test.go against our wired API server to ensure our default authz/admission set up protects us.

File: vendor/k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go [hunk 5]
```diff
@@ -405,10 +405,12 @@ func ClusterRoleBindings() []rbac.ClusterRoleBinding {
 		rbac.NewClusterBinding("system:kube-dns").SAs("kube-system", "kube-dns").BindingOrDie(),
 		rbac.NewClusterBinding("system:kube-scheduler").Users(user.KubeScheduler).BindingOrDie(),
 
-		// This default system:nodes binding is deprecated in 1.7 with the availability of the Node authorizer.
-		// If an admin wants to grant the system:node role (which cannot partition Node API access), they will need to create their own clusterrolebinding.
-		// TODO: Remove the subjects from this binding in 1.8 (leave the empty binding for tightening reconciliation), and remove AddClusterRoleBindingFilter()
-		rbac.NewClusterBinding(systemNodeRoleName).Groups(user.NodesGroup).BindingOrDie(),
+		// This default binding of the system:node role to the system:nodes group is deprecated in 1.7 with the availability of the Node authorizer.
```
Comment (enj) [inline vendor/k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go]: Uhh upstream commit?

Comment (liggitt) [reply vendor/k8s.io/kubernetes/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go]: Yes, pushed too fast. Will fix

**Comment by liggitt** [comment]
comments addressed, PTAL
